document-store/src/main/java/org/hypertrace/core/documentstore/DatastoreProvider.java

```diff
@@ -1,45 +1,56 @@
 package org.hypertrace.core.documentstore;
 
+import static java.util.Map.entry;
+
 import com.typesafe.config.Config;
-import java.lang.reflect.Constructor;
 import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
+import java.util.Optional;
+import java.util.function.Supplier;
+import lombok.NonNull;
+import org.hypertrace.core.documentstore.model.config.ConnectionConfig;
+import org.hypertrace.core.documentstore.model.config.DatabaseType;
 import org.hypertrace.core.documentstore.mongo.MongoDatastore;
 import org.hypertrace.core.documentstore.postgres.PostgresDatastore;
 
 public class DatastoreProvider {
-
-  private static Map<String, Class<? extends Datastore>> registry = new ConcurrentHashMap<>();
-
-  static {
-    DatastoreProvider.register("Mongo", MongoDatastore.class);
-    DatastoreProvider.register("Postgres", PostgresDatastore.class);
-  }
+  private static final Map<DatabaseType, Supplier<Datastore>> DATASTORE_MAPPING =
+      Map.ofEntries(
+          entry(DatabaseType.MONGO, MongoDatastore::new),
+          entry(DatabaseType.POSTGRES, PostgresDatastore::new));
 
   /**
-   * Creates a DocDatastore, currently it creates a new client/connection on every invocation. We
-   * might add pooling later
-   *
-   * @return {@link Datastore}
+   * @deprecated Use {@link
+   *     DatastoreProvider#getDatastore(org.hypertrace.core.documentstore.model.config.ConnectionConfig)}
+   *     instead
    */
+  @Deprecated
   public static Datastore getDatastore(String type, Config config) {
+    return Optional.ofNullable(type)
+        .map(DatabaseType::getType)
+        .map(DATASTORE_MAPPING::get)
+        .map(Supplier::get)
+        .map(store -> initialize(store, config))
+        .orElseThrow(() -> new IllegalArgumentException("Unknown database type: " + type));
+  }
 
-    Class<? extends Datastore> clazz = registry.get(type.toLowerCase());
-    try {
-      Constructor<? extends Datastore> constructor = clazz.getConstructor();
-      Datastore instance = constructor.newInstance();
-      instance.init(config);
-      return instance;
-    } catch (Exception e) {
-      throw new IllegalArgumentException("Exception creating DocDatastore", e);
-    }
+  @SuppressWarnings("unused")
```

Comment:
Main motivation here is to move away from `typesafe` Config?

Reply:
Yup. Make it a first-class configuration, which can be generated from any source.
The type-safe to ConnectionConfig conversion is provided via a Utility class.